src/common/hist_util.h

```diff
@@ -201,62 +248,121 @@ class GHistIndexBlockMatrix {
+  }
+
+  inline void Add(const tree::GradStats& b) {
+    sum_grad += b.sum_grad;
+    sum_hess += b.sum_hess;
+  }
+
+  inline void Add(GradientPair p) {
+    this->Add(p.GetGrad(), p.GetHess());
+  }
+
+  inline void Add(GradType grad, GradType hess) {
+    sum_grad += grad;
+    sum_hess += hess;
+  }
+
+  inline tree::GradStats ToGradStat() const {
+    return tree::GradStats(sum_grad, sum_hess);
+  }
+
+  inline void SetSubstract(const GradStatHist& a, const GradStatHist& b) {
+    sum_grad = a.sum_grad - b.sum_grad;
+    sum_hess = a.sum_hess - b.sum_hess;
+  }
+
+  inline void SetSubstract(const tree::GradStats& a, const GradStatHist& b) {
+    sum_grad = a.sum_grad - b.sum_grad;
+    sum_hess = a.sum_hess - b.sum_hess;
+  }
+
+  inline GradType GetGrad() const { return sum_grad; }
+  inline GradType GetHess() const { return sum_hess; }
+  inline static void Reduce(GradStatHist& a, const GradStatHist& b) { // NOLINT(*)
+    a.Add(b);
+  }
+};
+
+using GHistRow = Span<GradStatHist>;
 
 /*!
  * \brief histogram of gradient statistics for multiple nodes
  */
 class HistCollection {
  public:
   // access histogram for i-th node
-  GHistRow operator[](bst_uint nid) const {
-    constexpr uint32_t kMax = std::numeric_limits<uint32_t>::max();
-    CHECK_NE(row_ptr_[nid], kMax);
-    tree::GradStats* ptr =
-        const_cast<tree::GradStats*>(dmlc::BeginPtr(data_) + row_ptr_[nid]);
-    return {ptr, nbins_};
+  inline GHistRow operator[](bst_uint nid) {
+    if (nid >= data_arr_.size()) {
+      AddHistRow(nid);
+    }
+    return { const_cast<GradStatHist*>(dmlc::BeginPtr(*data_arr_[nid])), nbins_};
   }
 
   // have we computed a histogram for i-th node?
-  bool RowExists(bst_uint nid) const {
-    const uint32_t k_max = std::numeric_limits<uint32_t>::max();
-    return (nid < row_ptr_.size() && row_ptr_[nid] != k_max);
+  inline bool RowExists(bst_uint nid) const {
+    return nid < data_arr_.size();
   }
 
   // initialize histogram collection
-  void Init(uint32_t nbins) {
-    nbins_ = nbins;
-    row_ptr_.clear();
-    data_.clear();
+  inline void Init(uint32_t nbins) {
+    if (nbins_ != nbins) {
+      for (size_t i = 0; i < data_arr_.size(); ++i) {
+        delete data_arr_[i];
+      }
+      data_arr_.clear();
+      nbins_ = nbins;
+    }
+  }
+
+  ~HistCollection() {
+    for (size_t i = 0; i < data_arr_.size(); ++i) {
+      delete data_arr_[i];
+    }
   }
 
   // create an empty histogram for i-th node
-  void AddHistRow(bst_uint nid) {
-    constexpr uint32_t kMax = std::numeric_limits<uint32_t>::max();
-    if (nid >= row_ptr_.size()) {
-      row_ptr_.resize(nid + 1, kMax);
+  inline void AddHistRow(bst_uint nid) {
+    if (data_arr_.size() <= nid) {
+      data_arr_.resize(nid + 1, nullptr);
     }
-    CHECK_EQ(row_ptr_[nid], kMax);
 
-    row_ptr_[nid] = data_.size();
-    data_.resize(data_.size() + nbins_);
+    if (data_arr_[nid] == nullptr) {
+      data_arr_[nid] = new std::vector<GradStatHist>;
```

Comment:
Why do you need to use new?  I would try to avoid raw new and delete all together.